### PR TITLE
Using Tapis Token for file api authentication

### PR DESCRIPTION
**Pull request by fnets**
## Overview

Securing the data endpoints that the reports on A2CPS uses for reports

## Related

* [[WP-4](https://jira.tacc.utexas.edu/browse/WP-4)]

## Changes

Added the functionality to pull out a user's sessionid and use it to get their tapis token. Then used that tapis token to authenticate their api pulls of data

## Testing

Test by going to: https://dev.a2cps.tacc.utexas.edu/reports/datastore-api-demo-latest/ 

and making sure data loads correctly

### CODE REVIEW

File: src/data_loading.py [hunk 1]
```diff
@@ -332,36 +335,82 @@ def get_api_blood_data(api_root = 'https://api.a2cps.org/files/v2/download/publi
...
+    print(vbr_api_root)
+    try:
+        response = requests.get(vbr_api_root + '/status/auth/', tapis_token)
+        print('vbr api response:')
+        print(response.json()['status'])
+        if response.json()['status'] == 'OK':
+            return True
+        else:
+            print("Unauthorized to access data")
+            raise Exception
```
Comment (nathanfranklin) [inline src/data_loading.py:411]: would you want to return False here.  Looks like the user of get_auth_status is expecting this method to return a boolean
```suggestion
            return False
```

File: src/data_loading.py [hunk 2]
```diff
@@ -332,36 +335,82 @@ def get_api_blood_data(api_root = 'https://api.a2cps.org/files/v2/download/publi
...
+        print('portal api response:')
+        print(response)
+        if response:
+            tapis_token = response.json()['token']
+            return tapis_token
+        else:
+            print("Unauthorized to access tapis token")
+            raise Exception
+    except Exception as e:
+        return('portal api error: {}'.format(e))
```
Comment (nathanfranklin) [inline src/data_loading.py:395]:
```suggestion
        print('portal api error: {}'.format(e))
        raise e
```
the calling of this method (~line 344) assumes that it returns a token. so best to raise an exception here.

File: src/data_loading.py [hunk 3]
```diff
@@ -332,36 +335,82 @@ def get_api_blood_data(api_root = 'https://api.a2cps.org/files/v2/download/publi
...
+    
+def get_auth_status(vbr_api_root, tapis_token = None):
```
Comment (nathanfranklin) [inline src/data_loading.py:397]:
```suggestion
def get_auth_status(vbr_api_root, tapis_token):
```

File: src/data_loading.py [hunk 4]
```diff
@@ -332,36 +335,82 @@ def get_api_blood_data(api_root = 'https://api.a2cps.org/files/v2/download/publi
...
+        response = requests.get(vbr_api_root + '/status/auth/', tapis_token)
+        print('vbr api response:')
+        print(response.json()['status'])
+        if response.json()['status'] == 'OK':
+            return True
+        else:
+            print("Unauthorized to access data")
+            raise Exception
+    except Exception as e:
+        return('vbr api error: {}'.format(e))
```
Comment (nathanfranklin) [inline src/data_loading.py:413]: The current use of get_auth_status is expecting this method to return a boolean.

Would you also want to log this and return a False?  or raise an exception?

Comment (nathanfranklin) [inline src/data_loading.py:413]: The current use of get_auth_status is expecting this method to return a boolean. So would you also want to log this and return a False?  or raise an exception?

```suggestion
        print('vbr api error: {}'.format(e))
        return False
```

File: src/data_loading.py [hunk 5]
```diff
@@ -503,8 +556,11 @@ def clean_adverse_events(adverse_events, consented, display_terms_dict_multi):
         # Convert numeric values to display values using dictionary
         for i in display_terms_dict_multi.keys():
             if i in multi_data.columns:
-                multi_data = multi_data.merge(display_terms_dict_multi[i], how='left', on=i)
-
+                try:
+                    multi_data = multi_data.merge(display_terms_dict_multi[i], how='left', on=i)
+                except Exception as e:
+                    print(display_terms_dict_multi[i])
+                    print('error: {}'.format(e))
```
Comment (nathanfranklin) [inline src/data_loading.py:563]: would you want a `raise e` here?

File: src/data_loading.py [hunk 6]
```diff
@@ -332,36 +335,82 @@ def get_api_blood_data(api_root = 'https://api.a2cps.org/files/v2/download/publi
...
+    print(coresessionid[:5])
+    tapis_token = get_tapis_token(portal_api_root, coresessionid)
+    print(tapis_token)
+    auth_status = get_auth_status(vbr_api_root, tapis_token)
+
+    if auth_status == True:
+        try:
+            # Load Json Data
+            subjects1_filepath = '/'.join([api_root,'subjects','subjects-1-latest.json'])
+            subjects1_request = requests.get(subjects1_filepath)
```
Comment (nathanfranklin) [inline src/data_loading.py:352]: I would have suspected that this call and the call on line 360 would require an auth (i.e. the bearer token). or am I mistaken there?  I know this is the last step of getting things to work so please ignore it till when appropriate.

File: src/data_loading.py (hunk 4 above)
Comment (fnets) [reply src/data_loading.py]: Good catch!

File: src/data_loading.py [hunk 7]
```diff
@@ -470,7 +515,10 @@ def create_clean_subjects(subjects_raw, screening_sites, display_terms_dict, dis
                 if subjects[i].dtype == np.float64:
                     # for display columns where data is numeric, merge on display dictionary, treating cols as floats to handle nas
                     display_terms[i] = display_terms[i].astype('float64')
-                subjects = subjects.merge(display_terms, how='left', on=i)
+                try:
+                    subjects = subjects.merge(display_terms, how='left', on=i)
+                except Exception as e:
+                    raise e
```
Comment (nathanfranklin) [inline src/data_loading.py:521]:
```suggestion
                subjects = subjects.merge(display_terms, how='left', on=i)
```

If nothing is done when catching and then re-raisign exception, then maybe best to not put in try/except.  Same applies to line 554.

**Review by neblinc — APPROVED** [review]
Overall good to go to handle security review.
Feel free to address comment or leave it for clean up PR.

File: src/app.py [hunk 8]
```diff
@@ -21,7 +17,7 @@
 
 
 # Path to Report files at TACC
-api_root = environ.get("API_ROOT") #'https://api.a2cps.org/files/v2/download/public/system/a2cps.storage.community/reports'
+files_api_root = environ.get("FILES_API_ROOT")
```
Comment (taoteg) [inline src/app.py:20]: Where is this variable used?

File: src/data_loading.py (hunk 7 above)
Comment (fnets) [reply src/data_loading.py:521]: I'll give it a try, this was originally put in to address an issue with the legacy code having a casting error.

File: src/data_loading.py [hunk 9]
```diff
@@ -196,172 +196,216 @@ def get_local_subjects_raw(data_directory):
...
-        }
-        return consort_data_json
+        except Exception as e:
+            traceback.print_exc()
+            return None
 
-    except Exception as e:
-        traceback.print_exc()
+    else:
+        print("Unauthorized attempt to access Consort data")
```
[the remark this answers is not in the document]
Comment (fnets) [reply src/data_loading.py:247]: More for logging in the pod. I'll switch these over to the logger tool.

**Comment by neblinc** [comment]
Closing pull request, this PR was merged into latest and then to main.

**Pull request closed without merging**